docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,185 @@
+
+3. **Evaluate index set**. Pick the top `n` indexes as the *initial recommended index set* and then launch the *Swap and Re-evaluate* algorithm. Note that executing the *Swap and Re-evaluate* algorithm is necessary as the `reduced_cost` of certain query may be **joint** effect of several indexes. Within this context, it is hard to independently estimate each index's contribution to the final `reduced_cost`. For example, assume there is an extremely slow query in input workload and the desired indexes for this query is `a` and `b`. However, the number of allowed recommended indexes for the whole workload is limited and for some reason, `a` ranks top `n` in the final score list while `b` is not. But there are chances that without `b`, `a` can no more optimize that extremely slow query. To tackle this problem, we have to *efficiently* enumerate possible optimal recommend index set for the workload. The difference from *random enumeration* is that we reasonably think the initial recommended index set composed of top `n` indexes in the score list is a *good start*. By swapping partial elements between the `recommend_set` and the `remaining_set`, we are trying *variates* of the *good start* recommend set in which we are more likely to get the real optimal recommend set as quickly as possible. The factor determining the quality of recommend index set is the overall cost of input workload after adding all indexes in recommend index set to the schema in memory. Following is snippets of pseudorandom code of `Swap and Re-evaluate` algorithm.  
+
+```golang
+// GCIS stands for Global Candidate Index Set
+sort(GCIS, GCIS.Score, Descend_Order)
+recommend_set = GCIS[:N]
+remaining_set = GCIS[N:]
+min_cost = EvaluateWorkload(workload, recommend_set)
+
+for {
+    v_recommend_set, v_remaining_set = SwapElements(recommend_set, remaining_set, swap_size)
+    variant_cost = EvaluateWorkload(workload, v_recommend_set)
+    if variant_cost < min_cost {
+        min_cost = variant_cost
+        recommend_set = v_recommend_set
+        remaining_set = v_remaining_set
+    }
+    ......
+    // loop until specific rounds or timeout, break
+}
+```
+----------------------------------------------
+### A quick exmaple for single-column index recommendation
+
+**Workload**:
+
+```
+- Query1: SELECT COUNT(*) FROM t1 where a = 1 and b = 5
+- Query2: SELECT a FROM t1 ORDER BY c limit 10 
+- Query3: DELETE FROM t1 WHERE a = 1 AND d = 1
+```
+where index on column *c* has already existed. 
+
+- **Step1**: recommend indexes for single query. For each query, compile it and get the `original_cost` output by SQL optimizer. Then add all possible hypothetic indexes *a, b, d* to the schema, compile the same query again. After the compiling, `LSIs` can be extracted from output execution plan as well as the `virtual_cost`. Denote `reduced_cost` as *original_cost - virtual_cost*. Assume this step outputs:
+
+```
+Query 1--LSIs: {a, b}, reduced_cost: 4
```

Comment:
Is it single column indices?

Reply:
Yes, two single column indices are recommended: *a* and *b*